Fix warnings in make for Ubuntu 20.04.2 LTS

### Description

Few warnings are getting appeared while make. 

#### How to reproduce

In  a clean instance, do make 

Or

Use make clean in existing instance and make

#### Screenshots

<img width="923" alt="warnings" src="https://user-images.githubusercontent.com/10155976/109613953-e805d680-7b57-11eb-98c4-983c544e348b.PNG">
<img width="838" alt="warnings1" src="https://user-images.githubusercontent.com/10155976/109613961-ea683080-7b57-11eb-9c8d-52db53acd42e.PNG">


### Versions

* Latest commiton master:
* Distributor ID:	Ubuntu
* Description:	Ubuntu 20.04.2 LTS
* Release:	20.04
* Codename:	focal

Hello @shaheemazmalmmd, 

I did some research and came up with the following observations/suggestions. Since suppressing/fixing warnings are critical to the application it is only wise to discuss before changing anything. 

### Warning 1
```bash
nomos.c: In function ‘list_dir’:
nomos.c:187:17: warning: passing argument 1 to restrict-qualified parameter aliases with argument 3 [-Wrestrict]
```
https://github.com/fossology/fossology/blob/ae63dbd1a7e237628843c08ad57f837717d539bc/src/nomos/agent/nomos.c#L187

*Reasons*: Appending `\n` to the `filename_buf` may result in overflow since we are writing the same buffer.
*Also From manpage*: https://linux.die.net/man/3/sprintf
![image](https://user-images.githubusercontent.com/74113200/110777893-aaa3f600-8287-11eb-95cf-9d47bc8cbd65.png)

*Possible Fix*: Use `snprintf` for fixed-size buffer 
*Reference*: https://stackoverflow.com/questions/51697753/how-to-suppress-sprintf-warning-directive-writing-between-1-and-11-bytes-into

#### Warning 2
```bash
In function ‘strncpy’,
    inlined from ‘main’ at nomos.c:314:5:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10: warning: ‘__builtin_strncpy’ specified bound 64 equals destination size [-Wstringop-truncation]
```
https://github.com/fossology/fossology/blob/ae63dbd1a7e237628843c08ad57f837717d539bc/src/nomos/agent/nomos.c#L314

*Reasons: Copying max `sizeof(gl.progName)` bytes to `gl.progName` doesn't leave space for `\0` char.
*Reference*: https://stackoverflow.com/questions/50198319/gcc-8-wstringop-truncation-what-is-the-good-practice

*Possible Fix*: Copy max `sizeof(gl.progName) - 1` bytes or suppress the warning if null termination is not desired or source size never reaches max size.

### Warning 3
```bash
In function ‘strncpy’,
    inlined from ‘createRelativePath’ at licenses.c:580:5,
    inlined from ‘createRelativePath’ at licenses.c:569:7:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10: warning: ‘__builtin_strncpy’ output may be truncated copying 4095 bytes from a string of length 4095 [-Wstringop-truncation]
```
https://github.com/fossology/fossology/blob/master/src/nomos/agent/licenses.c#L580-L582
*Reason* Copying the same size buffer may result in truncation of the source buffer. I also suspect it may overflow.
*Possible Fix*: If this is theoretically established that destination (`scp-fullpath`) never overflows, suppressing the warning is our best bet.

### Warning 4
```bash
n function ‘strncpy’,
    inlined from ‘printRegexMatch’ at util.c:1014:5:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10: warning: ‘__builtin_strncpy’ output may be truncated copying 255 bytes from a string of length 4095 [-Wstringop-truncation]
```
*Reasons+Possible Fix*: same as **Warning 3**

### Warning 5
```bash
nomos_utils.c: In function ‘add2license_ref’:
nomos_utils.c:63:18: warning: ‘%s’ directive writing up to 4095 bytes into a region of size 4041 [-Wformat-overflow=]
```
https://github.com/fossology/fossology/blob/ae63dbd1a7e237628843c08ad57f837717d539bc/src/nomos/agent/nomos_utils.c#L63
*Reasons*: Possible buffer overflow
*Possible Fix*: Using `snprintf` with calculated max-size to copy, can suppress if no overflow possible theoritically

### Warning 6
```bash
nomos_utils.c:79:93: warning: ‘%s’ directive writing up to 4095 bytes into a region of size 4023 [-Wformat-overflow=]
```
https://github.com/fossology/fossology/blob/ae63dbd1a7e237628843c08ad57f837717d539bc/src/nomos/agent/nomos_utils.c#L79

*Reasons+Possible Fix*: Same as *Warning 5*

### Warning 7
```bash
In function ‘strncpy’,
    inlined from ‘convertWhitespaceToSpaceAndRemoveSpecialChars’ at doctorBuffer_utils.c:382:16:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10: warning: ‘__builtin_strncpy’ output truncated copying 7 bytes from a string of length 10 [-Wstringop-truncation]
```
*Reasons+Possible Fix*: Same as **Warning 3** but Not significant, suppressing is ok

### Warning 8
```bash
event.c: In function ‘event_loop_take’:
event.c:103:3: warning: ‘GTimeVal’ is deprecated: Use 'GDateTime' instead [-Wdeprecated-declarations]
```
*Reasons*: `GTimeVal` is susceptible to the *Year 2038 problem* and I hope this project will not die that soon :)
*Reference*:https://developer.gnome.org/glib/stable/glib-Date-and-Time-Functions.html#GTimeVal
*Possible Fix*: Replace with `GDateTime`

### Warning 9 and further

*Reasons +Possible Fix*: Similar to **Warning 2** and other warnings


Please have a look and let me know what should be fixed/suppressed or of any ambiguity in my observations. I will start working on that.

--- a/src/nomos/agent/nomos.c
+++ b/src/nomos/agent/nomos.c
@@ -310,13 +310,13 @@ int main(int argc, char **argv)
   /* Record the progname name */
   if ((cp = strrchr(*argv, '/')) == NULL_STR)
   {
-    strncpy(gl.progName, *argv, sizeof(gl.progName));
+    strncpy(gl.progName, *argv, sizeof(gl.progName)-1);
   }
   else
   {
     while (*cp == '.' || *cp == '/')
       cp++;
-    strncpy(gl.progName, cp, sizeof(gl.progName));
+    strncpy(gl.progName, cp, sizeof(gl.progName)-1);
   }
 
   if (putenv("LANG=C") < 0)

--- a/src/wget_agent/agent/main.c
+++ b/src/wget_agent/agent/main.c
@@ -214,7 +214,7 @@ int main  (int argc, char *argv[])
   for(arg=optind; arg < argc; arg++)
   {
     memset(GlobalURL,'\0',sizeof(GlobalURL));
-    strncpy(GlobalURL,argv[arg],sizeof(GlobalURL));
+    strncpy(GlobalURL,argv[arg],sizeof(GlobalURL)-1);
     /* If the file contains "://" then assume it is a URL.
        Else, assume it is a file. */
     LOG_VERBOSE0("Command-line: %s",GlobalURL);

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -1058,7 +1058,7 @@ void replace_url_with_auth()
       token = strtok(NULL, needle);
       index++;
     }
-    snprintf(GlobalURL, URLMAX, "%s%s:%s@%s", http, username, password, URI);
+    snprintf(GlobalURL, URLMAX-1, "%s%s:%s@%s", http, username, password, URI);
 
     if (strlen(additionalParams) > 0) {
       memmove(GlobalParam, additionalParams, strlen(additionalParams) +1);

